File: Source/Compiler/src/StringStream.cs
namespace Swift
{
	/*
	 *	Crappy implementation of StringStream.
	 *	Still less crappy than any default .NET implementations of StringStream.
	 *	Rewrite using TextReader
	 */

	internal class StringStream
	{
		public const char NewLine = '\n';

		private readonly string value;
		public int Position;

		public bool IsEndOfStream
		{
			get
			{
				return Position >= value.Length;
			}
		}

		public StringStream(string value)
		{
			this.value = value;
			Position = 0;
		}

		public char Peek(int offset = 0)
		{
			return Position + offset < value.Length ? value[Position + offset] : '\0';
		}

		public char Read()
		{
			return Position < value.Length ? value[Position++] : '\0';
		}
	}
}

File: Source/IDE/src/Project/Lexer/StringStream.cpp
/*
 *	StringStream.hpp
 *	Lexer's String Stream
 */

#include <Project/Lexer/StringStream.hpp>

namespace Lexer
{
	StringStream::StringStream(const QString& value)
		: value(value),
		  position(0)
	{

	}

	char StringStream::Peek(int offset)
	{
		return position + offset < value.length()
				? value[position + offset].toLatin1()
				: '\0';
	}

	char StringStream::Read()
	{
		return position < value.length()
			? value[position++].toLatin1()
			: '\0';
	}

	bool StringStream::IsEos() const
	{
		return position >= value.length();
	}
}
